paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java

```diff
@@ -238,32 +285,54 @@ private boolean isEmptySequenceGroup(KeyValue kv, FieldsComparator comparator) {
 
     private void retractWithSequenceGroup(KeyValue kv) {
         Set<Integer> updatedSequenceFields = new HashSet<>();
+        Iterator<WrapperWithFieldIndex<FieldsComparator>> comparatorIter =
+                fieldSeqComparators.iterator();
+        WrapperWithFieldIndex<FieldsComparator> curComparator =
+                comparatorIter.hasNext() ? comparatorIter.next() : null;
+        Iterator<WrapperWithFieldIndex<FieldAggregator>> aggIter = fieldAggregators.iterator();
+        WrapperWithFieldIndex<FieldAggregator> curAgg = aggIter.hasNext() ? aggIter.next() : null;
+
+        GenericRow currentValueRow = new GenericRow(getters.length);
+        currentValueRow.setRowKind(kv.valueKind());
+        for (int i = 0; i < getters.length; i++) {
+            currentValueRow.setField(i, getters[i].getFieldOrNull(kv.value()));
+        }
```

Comment:
This adds additional gets and sets to a Row object compared with the original implementation, which might be a potential performance regression.

How about adding a CachingRow implementation that caches the field values of a wrapped InternalRow into an array? This could help avoid invoking all getters at the beginning, while also avoiding invoking field getters multiple times on the same field.